### PR TITLE
Fix `dhall-json` for new `Prelude.JSON.Type`

### PR DESCRIPTION
**Pull request by Gabriel439**
... based on an issue caught by @NicolasT

https://github.com/dhall-lang/dhall-lang/pull/873 changed
`Prelude.JSON.Type` which requires matching changes to `dhall-json`
to support encoding/decoding Dhall expressions of that type.  We
still support the old type, too.

### CODE REVIEW

File: dhall-json/src/Dhall/JSONToDhall.hs [hunk 1]
```diff
@@ -590,6 +590,77 @@ dhallFromJSON (Conversion {..}) expressionType =
 
           return result
 
+    -- Arbitrary JSON ~> https://prelude.dhall-lang.org/JSON/Type (v13.0.0 <=)
```
Comment (sjakobi) [inline dhall-json/src/Dhall/JSONToDhall.hs]: Nice! I hadn't realized that `json-to-dhall` supports this at all!

File: dhall-json/tasty/data/fromArbitraryJSON_12.0.0.dhall [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      , object : List { mapKey : Text, mapValue : JSON } → JSON
+      , string : Text → JSON
+      }
+    )
+→ json.object
+    [ { mapKey = "string", mapValue = json.string "ABC" }
+    , { mapKey = "null", mapValue = json.null }
+    , { mapKey = "array", mapValue = json.array ([] : List JSON) }
+    , { mapKey = "number", mapValue = json.number 1.0 }
+    , { mapKey = "bool", mapValue = json.bool False }
```
Comment (sjakobi) [inline dhall-json/tasty/data/fromArbitraryJSON_12.0.0.dhall]: Is this order of `mapKey`s defined by `json-to-dhall`?

I think ideally we'd preserve the order of keys in the JSON document. The next best thing would IMHO be to use ascending lexicographic ordering.

But that's something for another ticket.

File: dhall-json/tasty/data/fromArbitraryJSON_13.0.0.dhall [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      , object : List { mapKey : Text, mapValue : JSON } → JSON
+      , string : Text → JSON
+      }
+    )
+→ json.object
+    [ { mapKey = "double", mapValue = json.double 1.0 }
+    , { mapKey = "string", mapValue = json.string "ABC" }
+    , { mapKey = "null", mapValue = json.null }
+    , { mapKey = "array", mapValue = json.array ([] : List JSON) }
+    , { mapKey = "integer", mapValue = json.double 1.0 }
```
Comment (sjakobi) [inline dhall-json/tasty/data/fromArbitraryJSON_13.0.0.dhall]: I'd expect a `json.integer` `mapValue` here?!

**Comment by sjakobi** [comment]
Regarding the Hydra failure, I suspect you need to enhance some `extra-src-files`.

File: dhall-json/tasty/data/fromArbitraryJSON_12.0.0.dhall (hunk 2 above)
Comment (Gabriel439) [reply dhall-json/tasty/data/fromArbitraryJSON_12.0.0.dhall]: It's the order of keys in the `HashMap` stored inside the `Object` constructor of `Value`.  The `HashMap` does not preserve order (one of the deficiencies of the `aeson` package), so you don't get them in some hash-dependent order.

That said, I could include a fix in this pull request to sort them.

File: dhall-json/src/Dhall/JSONToDhall.hs (hunk 1 above)
Comment (Gabriel439) [reply dhall-json/src/Dhall/JSONToDhall.hs]: Yeah, this is documented here: https://github.com/Gabriel439/dhall-manual/blob/a0b4ceb9fead3ba151fc28f3b389573493b66f6b/manuscript/01-YamlToDhall.md

File: dhall-json/tasty/data/fromArbitraryJSON_12.0.0.dhall (hunk 2 above)
Comment (sjakobi) [reply dhall-json/tasty/data/fromArbitraryJSON_12.0.0.dhall]: > That said, I could include a fix in this pull request to sort them.

That seems reasonable to me, but I don't mind too much.

**Pull request merged**